### ISSUE TITLE
Type variable in parameter cannot be returned with the same type

### ISSUE THREAD

**Issue by Bluenix2** [microsoft/pyright#3374]
## Summary

I have a function typed as `(cls: Type[T] = str) -> T` but simply returning `cast(cls, ...)` fails.

## Reproducible code example

```python
from typing import Type, TypeVar, cast

T = TypeVar('T')


def example(cls: Type[T] = str) -> T:
    return cast(cls, ...)
```

### Pyright error

```
Expression of type "T@example | str" cannot be assigned to return type "T@example"
  Type "T@example | str" cannot be assigned to type "T@example"
```

## Installation information
```
Released on   22/03/2019, 04:10:23
Last updated  20/04/2022, 07:27:55
Identifier    ms-pyright.pyright
```

**Comment by erictraut** [comment]
Yes, this is intended behavior. This is a type violation, so it's appropriate for pyright to emit an error here.

When you call `cast`, the first argument must be a type, but your'e passing the parameter `cls`. This parameter has the type `Type[T] | Type[str]`. The first part of the union comes from the explicit parameter annotation, and the second part comes from the default argument.

Note that mypy disallows both the use of a default argument for a generic parameter and the use of `cls` in the call to `cast`. Pyright allows both of these, but it (correctly) flags the type violation for the return type.

I can't tell exactly what you're trying to do here because your sample doesn't include much context, but I think what you probably want to do in this case is use `cast(T, ...)`.

**Comment by Bluenix2** [comment]
I see, yes I was looking for `cast(T, ...)` thank you. In my real code I wrap a `ContextVar` since subclassing the type of the value it contains is common.

I did notice that the type became `Type[T] | Type[str]`, can't Pyright infer `T = str` in this situation? I see my mistake with `cast()` but take this slightly modified example:

```python
from typing import Type, TypeVar, cast

T = TypeVar('T')


def instantiate(cls: Type[T] = str) -> T:
    return cls()

reveal_type(instantiate())
```
```
Expression of type "T@instantiate | str" cannot be assigned to return type "T@instantiate"
  Type "T@instantiate | str" cannot be assigned to type "T@instantiate"
```
```
Type of "instantiate()" is "str"
```

**Comment by erictraut** [comment]
Other Python type checkers don't allow default arguments to be used with parameters annotated with a TypeVar. The same is true of other languages like TypeScript, Swift and C#. Maybe it would be best if I simply disallowed it in pyright as well. After all, `str` is not assignable to `Type[T]` within the function, so it should arguably not be assignable as a default argument.

Refer to [this thread](https://github.com/microsoft/pyright/issues/3130) and the threads it links to for additional details.
